src/sentry/similarity.py

```diff
@@ -272,3 +297,209 @@ def record(self, scope, key, characteristics):
+        except KeyError as error:
+            self.logger.info('Could not create signature(s) for %r due error: %r', event, error, exc_info=True)
+            return
+
+        for exception in exceptions:
+            try:
+                yield self.function(exception)
+            except Exception as error:
+                self.logger.exception('Could not create signature for exception in %r due to error: %r', event, error)
+
+
+class MessageProcessor(object):
+    def __init__(self, function):
+        self.function = function
+        self.logger = logging.getLogger(__name__)
+
+    def process(self, event):
+        try:
+            message = event.data['sentry.interfaces.Message']
+        except KeyError as error:
+            self.logger.info('Could not create signature(s) for %r due error: %r', event, error, exc_info=True)
+            return
+
+        try:
+            yield self.function(message)
+        except Exception as error:
+            self.logger.exception('Could not create signature for message of %r due to error: %r', event, error)
+
+
+class ProcessorSet(object):
+    def __init__(self, index, aliases, processors):
+        self.index = index
+        self.aliases = aliases
+        self.processors = processors
+        self.__number_format = get_number_format(0xFFFFFFFF)
+        assert set(self.aliases) == set(self.processors)
+
+    def record(self, event):
+        items = []
+        for label, processor in self.processors.items():
+            alias = self.aliases[label]
+            scope = ':'.join((
+                alias,
+                self.__number_format.pack(event.project_id),
+            ))
+            for characteristics in processor.process(event):
+                if characteristics:
+                    items.append((
+                        scope,
+                        self.__number_format.pack(event.group_id),
+                        characteristics,
+                    ))
+        return self.index.record_multi(items)
+
+    def query(self, group):
+        results = {}
+        key = self.__number_format.pack(group.id)
+        for label in self.processors.keys():
+            alias = self.aliases[label]
+            scope = ':'.join((
+                alias,
+                self.__number_format.pack(group.project_id)
+            ))
+            results[label] = map(
+                lambda (id, score): (
+                    self.__number_format.unpack(id)[0],
+                    score,
+                ),
+                self.index.query(scope, key)
+            )
+        return results
+
+
+def serialize_text_shingle(value, separator=b''):
+    """\
+    Convert a sequence of Unicode strings into a bytes object.
+    """
+    return separator.join(
+        map(
+            operator.methodcaller('encode', 'utf8'),
+            value,
+        ),
+    )
+
+
+processors = ProcessorSet(
+    MinHashIndex(
+        redis.clusters.get('ephemeral'),
```

Comment:
This needs to be configurable by default (`ephemeral` is only relevant to our own usage), and probably moved to [`sentry.app` with other similar services.](https://github.com/getsentry/sentry/blob/master/src/sentry/app.py)